core/main/src/service/apps/apps_updater.rs

```diff
@@ -72,31 +211,67 @@ impl ExtnStreamProcessor for AppsUpdater {
 impl ExtnEventProcessor for AppsUpdater {
     async fn process_event(
         state: Self::STATE,
-        _msg: ExtnMessage,
+        _: ExtnMessage,
         extracted_message: Self::VALUE,
     ) -> Option<bool> {
-        if let AppsUpdate::AppsCatalogUpdate(apps_catalog_update) = extracted_message {
-            update(
-                state.client.clone(),
-                apps_catalog_update,
-                state.ignore_list,
-                state.uninstalls_enabled,
-            )
-            .await;
+        match extracted_message {
+            AppsUpdate::AppsCatalogUpdate(apps_catalog_update) => {
+                update(state, apps_catalog_update).await;
+            }
+            AppsUpdate::InstallComplete(op) => {
+                install_complete(state, op);
+            }
+            AppsUpdate::UninstallComplete(_) => {}
         }
         None
     }
 }
 
-pub async fn update(
-    mut client: ExtnClient,
-    apps_catalog_update: AppsCatalogUpdate,
-    ignore_list: Vec<String>,
-    uninstalls_enabled: bool,
-) {
-    debug!("update: apps_catalog_update={:?}", apps_catalog_update);
+/// Checks if the given installed app is
+/// the expected app to be installed
+/// If it is the same version in the old catalog
+/// or if it is the last_good_version before a failed install
+fn is_expected_install(
+    old_catalog: &Option<Vec<app_catalog::AppMetadata>>,
+    failed_installs: &Arc<RwLock<Vec<FailedAppInstall>>>,
+    installed_app: &InstalledApp,
+) -> Option<bool> {
+    let in_old_catalog = old_catalog.as_ref().map(|apps| {
```

Comment:
I'm a little confused what the meaning of the `Option<bool>` return value is, can you explain or map it out in the function comment?

Reply:
We return None for when old_catalog is None. It basically means we don't know if this app is expected to be installed.
I guess technically based on the method name we should return false. If there is no old catalog, then NO apps should be installed. So the fact any app is installed means it is "not expected", but that breaks down when we go from softcat to distp. If this is the very first distp catalog we have installed, then the softcat version IS expected.
To make it simpler this method can return:
Yes it is expected, No it is not expected, I don't know if this is expected.
In the I don't know case we always choose to install the app from distp.